pkg/sql/unsplit_range_test.go

```diff
@@ -0,0 +1,460 @@
+}
+
+// splitLastRangeInSpan split the last range into two ranges,
+// and check the new range has sticky bit set.
+func splitLastRangeInSpan(
+	ctx context.Context, t *testing.T, kvDB *kv.DB, tableSpan roachpb.Span,
+) roachpb.Key {
+	metaStartKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.Key))
+	metaEndKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.EndKey))
+	ranges, err := kvDB.Scan(ctx, metaStartKey, metaEndKey, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lastRange := ranges[len(ranges)-1]
+	var lastRangeDesc roachpb.RangeDescriptor
+	if err := lastRange.ValueProto(&lastRangeDesc); err != nil {
+		t.Fatal(err)
+	}
+	splitKey := getMidKeyInSpan(t, kvDB, lastRangeDesc.StartKey, lastRangeDesc.EndKey)
+	splitKey, err = keys.EnsureSafeSplitKey(splitKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := kvDB.AdminSplit(ctx, splitKey, hlc.MaxTimestamp); err != nil {
+		t.Fatal(err)
+	}
+
+	return splitKey
+}
+
+// rangeIsManuallySplit check if there is a range starts with `startKey` and is
+// manually split
+func rangeIsManuallySplit(
+	ctx context.Context, t *testing.T, kvDB *kv.DB, tableSpan roachpb.Span, startKey roachpb.Key,
+) bool {
+	metaStartKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.Key))
+	metaEndKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.EndKey))
+	ranges, err := kvDB.Scan(ctx, metaStartKey, metaEndKey, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.NotEmpty(t, ranges)
+	for _, r := range ranges {
+		var desc roachpb.RangeDescriptor
+		if err := r.ValueProto(&desc); err != nil {
+			t.Fatal(err)
+		}
+		if bytes.Equal(desc.StartKey.AsRawKey(), startKey) && !desc.GetStickyBit().IsEmpty() {
+			return true
+		}
+	}
+	return false
+}
+
+// TODO(Chengxiong): remove test for old version and update this comment.
+// Test that manually split ranges get unsplit when dropping a table/database.
+// It verifies that the logic is working on both the old (before version
+// `UnsplitRangesInAsyncGCJobs`) and new (from versiom
+// `UnsplitRangesInAsyncGCJobs`) pathes.
+func TestRangesUnsplitWhenTableDropped(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	testCases := []struct {
+		name          string
+		query         string
+		binaryVersion clusterversion.Key
+	}{
+		{
+			name:          "drop-table-unsplit-sync",
+			query:         "DROP TABLE t.test1",
+			binaryVersion: clusterversion.UnsplitRangesInAsyncGCJobs - 1,
+		},
+		{
+			name:          "drop-table-unsplit-async",
+			query:         "DROP TABLE t.test1",
+			binaryVersion: clusterversion.UnsplitRangesInAsyncGCJobs,
+		},
+		{
+			name:          "drop-database-unsplit-sync",
+			query:         "DROP DATABASE t",
+			binaryVersion: clusterversion.UnsplitRangesInAsyncGCJobs - 1,
+		},
+		{
+			name:          "drop-database-unsplit-async",
+			query:         "DROP DATABASE t",
+			binaryVersion: clusterversion.UnsplitRangesInAsyncGCJobs,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
```

Comment:
these tests still share a very large amount of structure. Can you rework the cases to plumb through a few more arguments into the test case so that all of these can be subtests of a single test. I really don't think there's much to do to make that happen.